src/main/java/com/axis/system/jenkins/plugins/downstream/cache/BuildCache.java

```diff
@@ -182,46 +209,82 @@ protected void updateCache(Run downstreamRun) {
   }
 
   public void doGarbageCollect() {
+    Logger.info("Running GC...");
     try (ACLContext ignored = ACL.as(ACL.SYSTEM)) {
       downstreamBuildCache
           .entrySet()
           .removeIf(
               e -> {
-                if (Run.fromExternalizableId(e.getKey()) == null) {
-                  LOGGER.info(e.getKey() + " will be GC:ed");
+                // Stop looking up builds if we are trying to shutdown.
+                if (!workerThreadPool.isShutdown()
+                    && Run.fromExternalizableId(e.getKey()) == null) {
+                  Logger.info(e.getKey() + " will be GC:ed");
                   return true;
                 }
                 return false;
               });
     }
+    Logger.info("GC completed!");
   }
 
-  public synchronized void stopGarbageCollector() {
-    LOGGER.info("Stopping GC scheduling");
-    if (timer != null) {
-      timer.cancel();
-    }
-  }
-
-  public synchronized void startGarbageCollector() {
-    LOGGER.info("Setting up GC scheduling");
-    if (timer == null) {
-      timer = new Timer();
-    } else {
-      timer.cancel();
-      timer = new Timer();
+  public void setupGarbageCollector() {
+    Logger.info("Setting up GC scheduling");
+    if (gcTimer != null) {
+      gcTimer.cancel();
     }
-    timer.scheduleAtFixedRate(
+    gcTimer = new Timer();
+    gcTimer.scheduleAtFixedRate(
         new TimerTask() {
           @Override
           public void run() {
-            doGarbageCollect();
+            workerThreadPool.execute(BuildCache.this::doGarbageCollect);
```

Comment:
May I suggest a guard:
if (!isCacheRefreshing()) {
   WorkerThreadPool.execute(...)
}
that way we don't add loads of cleanup tasks while we are reloading cache.

Reply:
Agree